### PR TITLE
removing Rounds and adding its properties to PoseidonParameters

### PR DESCRIPTION
**Pull request by mfaghihi**
WIP

### CODE REVIEW

File: src/merkle_tree/mod.rs [hunk 1]
```diff
@@ -452,9 +446,21 @@ mod test {
...
+		let width_3 = get_width_poseidon_bls381_x5_3::<Fq>();
+		let sbox_3 = get_sbox_poseidon_bls381_x5_3::<Fq>();
+		let params3 = PoseidonParameters::<Fq>::new(
+			round_keys_3,
+			mds_matrix_3,
+			full_rounds_3,
+			partial_rounds_3,
+			width_3,
+			sbox_3,
+		);
```
Comment (filiplazovic) [inline src/merkle_tree/mod.rs:462]: All of these functions should be moved to the files where constants are defined + new function should be added for creating new PoseidonParameters instance: i.e.
```
fn get_poseidon_bls381_x5_3<F: PrimeField>() -> PoseidonParameters<F> {
  let rounds = get_rounds_poseidon_bls381_x5_3();
  let mds = get_mds_poseidon_bls381_x5_3;
  PoseidonParameters::<Fq>::new(
	rounds,
	mds,
	FULL_ROUND,
	PARTIAL_ROUND,
	WIDTH,
	SBOX,
  )
}
```

Comment (mfaghihi) [reply src/merkle_tree/mod.rs:462]: Good point. We already have some functions that to this job. We should just use them.

File: src/utils/bls381_x17_3.rs [hunk 2]
```diff
@@ -18,6 +18,41 @@ pub const PARTIAL_ROUNDS: u8 = 57;
...
+pub fn get_mds_poseidon_bls381_x17_3<F: PrimeField>() -> Vec<Vec<F>> {
+	parse_matrix(
+		MDS_ENTRIES
+			.iter()
+			.map(|x| x.to_vec())
+			.collect::<Vec<_>>(),
+	)
+}
+
+#[cfg(feature = "poseidon_bls381_x17_3")]
```
Comment (filiplazovic) [inline src/utils/bls381_x17_3.rs:38]: No need for this feature gate, it already applied for the whole file

File: src/merkle_tree/mod.rs (hunk 1 above)
Comment (filiplazovic) [reply src/merkle_tree/mod.rs]: I think this is the only thing left. So move these https://github.com/webb-tools/arkworks-gadgets/blob/hossein/poseidon_modification/src/setup/common.rs#L206 into the file where constants are defined

Comment (mfaghihi) [reply src/merkle_tree/mod.rs]: It looks more organized If don't move them.

**Review by drewstone — APPROVED** [review]
LGTM

File: src/poseidon/mod.rs [hunk 3]
```diff
@@ -106,6 +111,23 @@ impl<F: PrimeField> PoseidonParameters<F> {
 	}
 
 	pub fn from_bytes(mut bytes: &[u8]) -> Result<Self, Error> {
+		let mut width_u8 = [0u8; 8];
```
Comment (filiplazovic) [inline src/poseidon/mod.rs:114]: Let's use `usize::BITS / 8` since on some machines usize could be 32 bits

File: src/poseidon/mod.rs [hunk 4]
```diff
@@ -273,75 +300,50 @@ mod test {
...
+		   let partial_rounds_3 = get_partial_rounds_poseidon_bn254_x5_3::<Fq>();
+		   let width_3 = get_width_poseidon_bn254_x5_3::<Fq>();
+		   let sbox_3 = get_sbox_poseidon_bn254_x5_3::<Fq>();
+		   let	params = PoseidonParameters::<Fq>::new(round_keys_3, mds_matrix_3, full_rounds_3, partial_rounds_3, width_3, sbox_3);
+
+		   let bytes = params.to_bytes();
+		   let new_params: PoseidonParameters<Fq> = PoseidonParameters::from_bytes(&bytes).unwrap();
+		   assert_eq!(bytes, new_params.to_bytes());
+	   }
+	*/
```
Comment (filiplazovic) [inline src/poseidon/mod.rs:325]: Lets fix this test as well

File: src/utils/bn254_x5_2.rs [hunk 5]
```diff
@@ -138,3 +162,14 @@ pub const MDS_ENTRIES: [[&str; 2]; 2] = [
...
+/*
+#[derive(Default, Clone)]
+struct PoseidonRounds2;
+
+impl Rounds for PoseidonRounds2 {
+	const FULL_ROUNDS: u8 = 8;
+	const PARTIAL_ROUNDS: u8 = 56;
+	const SBOX: PoseidonSbox = PoseidonSbox::Exponentiation(5);
+	const WIDTH: u8 = 2;
+} */
```
Comment (filiplazovic) [inline src/utils/bn254_x5_2.rs:175]: Remove this

**Comment by filiplazovic** [comment]
This is good to go after the comments are addressed

File: src/poseidon/mod.rs [hunk 6]
```diff
@@ -38,24 +38,14 @@ impl ArkError for PoseidonError {}
...
-} */
-
 impl<F: PrimeField> PoseidonParameters<F> {
 	pub fn new(
 		round_keys: Vec<F>,
 		mds_matrix: Vec<Vec<F>>,
 		full_rounds: u8,
 		partial_rounds: u8,
 		width: u8,
+		exponentiation: u8,
```
Comment (filiplazovic) [inline src/poseidon/mod.rs:48]: Why is this added?

Comment (mfaghihi) [reply src/poseidon/mod.rs:48]: it stores the value for exponentiation for the sbox. it is used in the to_and_from_byte functions.  Is there a way to get the exponentiation from the sbox?

Comment (filiplazovic) [reply src/poseidon/mod.rs:48]:
```rust
let bytes = match sbox {
  Inverse => (0 as usize).to_bytes() // or similar
  Exponentiation(ex) => ex.to_bytes()
}
```

Comment (mfaghihi) [reply src/poseidon/mod.rs:48]: ok
